documentation/MSBuild-scheduler.md

```diff
@@ -0,0 +1,171 @@
+
+### BuildRequest
+Basic representation of build operation
+- `GlobalRequestId`: A unique identifier for the request across the entire build session. This ID stays constant regardless of which node processes the request. It allows the scheduler to track and reference the request throughout the build lifecycle. If two requests are semantically identical (same targets for the same project **configuration**), they may share the same GlobalRequestId to enable result reuse.
+- `ConfigurationId`: Identifies which project configuration this request is for. A configuration represents a unique combination of project file + global properties + tools version.
+- `NodeRequestId`: A node-specific identifier for the request. Each node maintains its own request IDs, which are used for local tracking on that node (whether in-process or out-of-process).
+- `SubmissionId`: an identifier that groups related BuildRequests that originate from a single top-level build command or operation.
+- `ParentGlobalRequestId`: References the GlobalRequestId of the request that created this one. For example, if ProjectA depends on ProjectB, the request to build ProjectB would have ProjectA's request ID as its ParentGlobalRequestId. This allows the scheduler to track dependencies and manage the build process effectively.
+- `Targets`: The list of targets to build in this project. These are the specific build tasks to execute, like "Clean", "Build", or "Publish".
+
+## Caching System
+
+### ConfigCache
+Project configuration storage
+- Project files
+- Global properties
+- Targets
+- ToolsVersion
+
+### ResultsCache
+Build result storage
+- Target outputs
+- Build success/failure
+- Target-level granularity
+
+## Scheduling Algorithms
+
+So the default scheduling algorithm varies based on:
+
+### Single-Processor Case (MaxNodeCount = 1):
+Uses AssignUnscheduledRequestsFIFO (First-In-First-Out)
+
+### Multi-Processor Case (MaxNodeCount > 1):
+- First calls AssignUnscheduledRequestsByTraversalsFirst to prioritize traversal projects
+- Then calls AssignUnscheduledProxyBuildRequestsToInProcNode for proxy builds
+- Then chooses based on whether a valid scheduling plan exists:
+
+#### With Valid Plan:
+- For 2 nodes: AssignUnscheduledRequestsWithPlanByMostImmediateReferences
+- For 3+ nodes: AssignUnscheduledRequestsWithPlanByGreatestPlanTime
+
+#### Without Valid Plan:
+- AssignUnscheduledRequestsWithConfigurationCountLevelling
+
+A "valid plan" in MSBuild refers to historical build data that was collected from previous successful builds and stored for optimization purposes.
+A scheduling plan contains information about:
+
+1) Build Times: How long each project took to build
+2) Project References: Which projects reference other projects
+3) Configuration Maps: How project paths map to configuration IDs
+
+This data is saved at the end of a successful build, typically in a file with extension `.buildplan`, so it can be used to make better scheduling decisions in future builds.
```

Comment:
does this really happen? I see that some code for .buildplan files exists but I have never seen it

Reply:
maybe mention `MSBUILDENABLEBUILDPLAN` in the doc?